fetch.bs

```diff
@@ -4068,11 +4068,47 @@ Range Requests</cite>. [[HTTP-RANGE]] However, this is not widely supported by b
    <a for=request>referrer</a>, <a lt="url serializer">serialized</a> and <a>isomorphic encoded</a>,
    to <var>httpRequest</var>'s <a for=request>header list</a>.
 
-   <li><p>If the <i>CORS flag</i> is set, <var>httpRequest</var>'s <a for=request>method</a> is
-   neither `<code>GET</code>` nor `<code>HEAD</code>`, or <var>httpRequest</var>'s
-   <a for=request>mode</a> is "<code>websocket</code>", then <a for="header list">append</a>
-   `<code>Origin</code>`/the result of <a>serializing a request origin</a> with
-   <var>httpRequest</var>, to <var>httpRequest</var>'s <a for=request>header list</a>.
+   <li><p>Let <var>serializedOrigin</var> be the result of <a>serializing a request origin</a> with
+   <var>httpRequest</var>.
+
+   <li><p>If the <i>CORS flag</i> is set, or <var>httpRequest</var>'s <a for=request>mode</a> is
+   "<code>websocket</code>", then <a for="header list">append</a>
+   `<code>Origin</code>`/<var>serializedOrigin</var> to <var>httpRequest</var>'s
+   <a for=request>header list</a>.
+
+   <li>
+    <p>Otherwise, if <var>httpRequest</var>'s <a for=request>method</a> is neither
+    `<code>GET</code>` nor `<code>HEAD</code>`, then switch on <var>httpRequest</var>'s
+    <a for=request>referrer policy</a>:
+
+    <dl class=switch>
+     <dt>"<code>no-referrer</code>"
+     <dd><p><a for="header list">Append</a> `<code>Origin</code>`/`<code>null</code>` to
+     <var>httpRequest</var>'s <a for=request>header list</a>.
+
+     <dt>"<code>no-referrer-when-downgrade</code>"
+     <dt>"<code>strict-origin</code>"
+     <dt>"<code>strict-origin-when-cross-origin</code>"
+     <dd><p>If <var>request</var>'s <a for=request>origin</a>'s <var>scheme</var> is
```

Comment:
While reformatting I realized that this cannot work as there's no guarantee request's origin is an origin that has a scheme. So this also needs to account for the origin being an opaque origin.

Another problem here is that if this conditional isn't met, no header gets appended. That seems unlikely to be the intention?

One thing I was thinking that might make this easier if we split this out into its own algorithm, perhaps right below "Serializing a request origin". Something like "Appending a request Origin header".

As I know you're somewhat otherwise occupied let me know if I should help out here.

Reply:
> While reformatting I realized that this cannot work as there's no guarantee request's origin is an origin that has a scheme. So this also needs to account for the origin being an opaque origin.

Done. Not sure if the fetch spec provide short-circuit behavior for (and/or)
> 
> Another problem here is that if this conditional isn't met, no header gets appended. That seems unlikely to be the intention?

It's horrible. Thanks for picking this.
> 
> One thing I was thinking that might make this easier if we split this out into its own algorithm, perhaps right below "Serializing a request origin". Something like "Appending a request Origin header".
> 

It's a bit complicated now, so IMO it's a good idea to isolate it.

@annevk Please take a look for the new patch. Thanks!